typing/includecore.ml

```diff
@@ -233,47 +286,86 @@ let report_constructor_mismatch first second decl ppf err =
         (String.capitalize_ascii (choose ord first second))
         (choose_other ord first second)
 
-let report_variant_mismatch first second decl ppf err =
+let report_variant_mismatch first second decl env ppf err =
   let pr fmt = Format.fprintf ppf fmt in
   match (err : variant_mismatch) with
   | Constructor_mismatch (c1, c2, err) ->
       pr
-        "@[<hv>Constructors do not match:@;<1 2>%a@ is not compatible with:\
+        "@[<hv>Constructors do not match:@;<1 2>%a@ is not the same as:\
          @;<1 2>%a@ %a@]"
         Printtyp.constructor c1
         Printtyp.constructor c2
-        (report_constructor_mismatch first second decl) err
+        (report_constructor_mismatch first second decl env) err
   | Constructor_names (n, name1, name2) ->
       pr "Constructors number %i have different names, %s and %s."
         n (Ident.name name1) (Ident.name name2)
   | Constructor_missing (ord, s) ->
       pr "The constructor %s is only present in %s %s."
         (Ident.name s) (choose ord first second) decl
 
-let report_extension_constructor_mismatch first second decl ppf err =
+let report_extension_constructor_mismatch first second decl env ppf err =
   let pr fmt = Format.fprintf ppf fmt in
   match (err : extension_constructor_mismatch) with
-  | Constructor_privacy -> pr "A private type would be revealed."
+  | Constructor_privacy ->
+      pr "Private extension constructor(s) would be revealed."
   | Constructor_mismatch (id, ext1, ext2, err) ->
-      pr "@[<hv>Constructors do not match:@;<1 2>%a@ is not compatible with:\
+      pr "@[<hv>Constructors do not match:@;<1 2>%a@ is not the same as:\
           @;<1 2>%a@ %a@]"
         (Printtyp.extension_only_constructor id) ext1
         (Printtyp.extension_only_constructor id) ext2
-        (report_constructor_mismatch first second decl) err
+        (report_constructor_mismatch first second decl env) err
+
+let report_private_variant_mismatch first second decl env ppf err =
+  let pr fmt = Format.fprintf ppf fmt in
+  match (err : private_variant_mismatch) with
+  | Only_outer_closed ->
+      (* It's only dangerous in one direction, so we don't have a position *)
+      pr "%s is private and closed, but %s is not closed"
+        (String.capitalize_ascii second) first
+  | Missing (ord, name) ->
+      pr "The constructor %s is only present in %s %s."
```

Comment:
I've added this everywhere I could (although this is squirrely enough that I may have missed some places); I did this as a separate commit, since it was very noisy but not very informative.  However, I'm wondering if the reverse is a better idea.  When we end with types, e.g. in `The type int is not compatible with the type bool -> M.t.`, it can be confusing to have the period.  But it's worse when there's a newline: I think you really don't want to include a period in

```
Type a is not compatible with type
  t = [ `Rack of 'a | `Rec of 'a | `Rock of 'a ] X.t as 'a
```

If you put it on the second line, it looks like it's part of the type, but you don't want to create a third line with just a period.  So what if we just banish periods entirely?  That avoids this potential confusion.  Currently, I intentionally don't add periods in any printing case that ends in a type, though of course this is easy enough to change.

Reply:
Sorry, I meant that we could keep this choice consistent across this specific PR, not across the whole code base. The last commit is impressive, and I will definitely review it if you submit it as a separate PR. However, I think it would be better to split this change to keep this PR more focused.

My personal opinion is that sentence ought to be punctuated (but maybe more with `bool -> M.t .` ). The potential confusion could be avoided with highlighting code block (like proposed in #10321). But that is probably my physics background speaking. Opening an issue might be the right way to probe for opinion more widely.

Reply:
Oh, silly me, of course you did :-)  I'll go back and make *this* PR consistent, yeah, of course.

(I share your desire to punctuate sentences, but I've always been torn between the mathematical practice of punctuating block quotes and the programming excerpt problem of not being able to punctuate code blocks.  Perhaps a space is the cleanest way?  I agree that opening an issue is the right call.)